src/requestor/routes/maintain.py

```diff
@@ -132,64 +129,72 @@ async def update_request(
     api_request: Request,
     request_id: uuid.UUID,
     status: str = Body(..., embed=True),
-    bearer_token: HTTPAuthorizationCredentials = Security(bearer),
+    auth=Depends(Auth),
 ) -> dict:
     """
     Update an access request with a new "status".
     """
-    request = await RequestModel.query.where(
-        RequestModel.request_id == request_id
-    ).gino.first_or_404()
-
-    await authorize(
-        api_request.app.arborist_client,
-        bearer_token,
-        "update",
-        [request.to_dict()["resource_path"]],
-    )
-
-    if request.status == status:
-        logger.debug(f"Request '{request_id}' already has status '{status}'")
-        return request.to_dict()
-
-    logger.debug(f"Updating request '{request_id}' with status '{status}'")
-
-    allowed_statuses = config["ALLOWED_REQUEST_STATUSES"]
-    if status not in allowed_statuses:
-        raise HTTPException(
-            HTTP_400_BAD_REQUEST,
-            f"Status '{status}' is not an allowed request status ({allowed_statuses})",
+    # only allow 1 update request at a time on the same row
+    async with db.transaction():
+        request = (
+            await RequestModel.query.where(RequestModel.request_id == request_id)
+            # lock the row by using FOR UPDATE clause
+            .with_for_update().gino.first_or_404()
```

Comment:
This makes a ton of sense and is a great addition. But two questions just since I'm unfamiliar with the issue and this endpoint/repo:

1. Just want to confirm that this should be a 404, right? It looks like this is for a PUT to an existing resource, but are there cases where the record for `request_id` may no longer be in the DB? That's why we're doing `first_or_404`?
2. Do we want to pair `wait_for_update` with `populate_existing` as specified in [this sqlalchemy note](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=update#sqlalchemy.orm.query.Query.with_for_update)? Not sure if that matters with our logic, but maybe a good practice if another transaction were to update this record while we were waiting for the lock?

Reply:
1. We return 404 if the `request_id` doesn't exist in the DB. Yes it should be a PUT to an existing resource but we can't trust users to only try to update things that exist :p
2. docs for reference: _"It is generally a good idea to combine the use of the Query.populate_existing() method when using the Query.with_for_update() method. The purpose of Query.populate_existing() is to force all the data read from the SELECT to be populated into the ORM objects returned, even if these objects are already in the identity map."_
The only case i can think of (this endpoint only allows updating `request.status`) where this could be useful is if the current `request.status` is "A", 2 concurrent requests try to update it to "B" and "A", the first one updates to "B" and the second one [doesn't update](https://github.com/uc-cdis/requestor/blob/6329a5af51304c2410b3df25d433d985ea3db0c4/src/requestor/routes/maintain.py#L152).
Error when using `.populate_existing()`: `AttributeError: 'Select' object has no attribute 'populate_existing'`
but from the doc i see `.execution_options(populate_existing=True)` is equivalent so i'll add that just in case

Reply:
😄 👍